### PR TITLE
[E098-MG < T0763-MG] Pulsar e2e

### PR DESCRIPTION
**Pull request by antonio2368**

### CODE REVIEW

File: tests/e2e/streams/common.py [hunk 1]
```diff
@@ -119,3 +128,152 @@ def check_stream_info(cursor, stream_name, expected_stream_info):
...
+    # This test is quite complex. The goal is to call START/STOP queries
+    # while a CHECK query is waiting for its result. Because the Global
+    # Interpreter Lock, running queries on multiple threads is not useful,
+    # because only one of them can call Cursor::execute at a time. Therefore
+    # multiple processes are used to execute the queries, because different
+    # processes have different GILs.
+    # The counter variables are thread- and process-safe variables to
+    # synchronize between the different processes. Each value represents a
+    # specific phase of the execution of the processes.
+    assert operation == "START" or operation == "STOP"
```
Comment (jbajic) [inline tests/e2e/streams/common.py:157]:
```suggestion
    assert operation in ["START", "STOP"]
```

File: tests/e2e/streams/pulsar_streams_tests.py [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+            common.pulsar_default_namespace_topic(topic),
+            send_timeout_millis=60000)
+        producer.send(common.SIMPLE_MSG)
+
+    for topic in pulsar_topics:
+        check_vertex_exists_with_topic_and_payload(
+            cursor, topic, common.SIMPLE_MSG)
+
+
+# @pytest.mark.skip(reason="no way of currently testing this")
```
Comment (jbajic) [inline tests/e2e/streams/pulsar_streams_tests.py:58]: Remove comment

**Review by jbajic — APPROVED** [review]

**Comment by jbajic** [comment]
Only a minor comment, overall looks good (we will standarize our py format and add precommit later on :tada: )

File: tests/e2e/streams/pulsar_streams_tests.py [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+        f"TOPICS {pulsar_topics[0]} "
+        f"TRANSFORM {transformation} "
+        "BATCH_SIZE 1",
+    )
+    common.start_stream(cursor, "test")
+    time.sleep(1)
+
+    producer = pulsar_client.create_producer(
+        common.pulsar_default_namespace_topic(
+            pulsar_topics[0]), send_timeout_millis=60000)
```
Comment (antaljanosbenjamin) [inline tests/e2e/streams/pulsar_streams_tests.py:186]: Can be the timeout be a bit smaller? I would say 10s should be plenty for this (and similar occasions), isn't it?

File: src/integrations/pulsar/consumer.cpp [hunk 4]
```diff
@@ -54,14 +55,16 @@ utils::BasicResult<std::string, std::vector<Message>> GetBatch(TConsumer &consum
...
     pulsar_client::Message message;
     const auto result = ConsumeMessage(consumer, message, remaining_timeout_in_ms);
     switch (result) {
       case pulsar_client::Result::ResultTimeout:
         return std::move(batch);
       case pulsar_client::Result::ResultOk:
-        batch.emplace_back(Message{std::move(message)});
+        if (message.getMessageId() != last_message_id) {
+          batch.emplace_back(Message{std::move(message)});
+        }
```
[the remark this answers is not in the document]
Comment (antonio2368) [reply src/integrations/pulsar/consumer.cpp:67]: Because the docs say that it should give the next message after the one you set but for some reason, it doesn't. I checked so many times the logs of Pulsar on client and server side and it doesn't make sense...
It starts from the message you send to the reader so this is a way of skipping it. Maybe it's just an edge case that happens in E2E tests but for now, I would leave it as safety.

File: tests/e2e/streams/pulsar_streams_tests.py (hunk 3 above)
Comment (antonio2368) [reply tests/e2e/streams/pulsar_streams_tests.py:186]: I set the same thing as for Kafka (unless it's not 60 seconds but a smaller unit...).

**Review by antaljanosbenjamin — APPROVED** [review]

File: src/query/procedure/mg_procedure_impl.cpp [hunk 5]
```diff
@@ -2496,16 +2497,9 @@ mgp_error mgp_message_payload(mgp_message *message, const char **result) {
...
-              using MessageType = std::decay_t<T>;
-              if constexpr (std::same_as<MessageType, mgp_message::KafkaMessage>) {
-                return msg->Payload().data();
-              } else if constexpr (std::same_as<MessageType, mgp_message::PulsarMessage>) {
-                return msg.Payload().data();
-              } else {
-                throw std::invalid_argument("Invalid source type");
-              }
-            },
+            utils::Overloaded{[](const mgp_message::KafkaMessage &msg) { return msg->Payload().data(); },
```
Comment (kostasrim) [inline src/query/procedure/mg_procedure_impl.cpp:2500]: :rocket: so compact and beautiful

**Review by kostasrim — APPROVED** [review]
Great job!
